docs/proposals/improve-addons.md

```diff
@@ -80,230 +84,184 @@ See below for a high level architecture view. Each component will be explained i
+- the ApplicationInstallationController would watch the cluster CR and create corresponding ApplicationInstallation CRs
 
-### CoreAddonController
+This implementation is close to the status quo of packing everything into the cluster object and then react to it. However we have the concern that the sum of all ApplicationInstallations with all of its fields can become quite large and result in a too large cluster object
 
-The CoreAddonController resides in the seed cluster. Its main purpose is to manage AddonInstallation CRs for CoreAddons. This job entails standard create/update/delete operations as well as picking the correct CoreAddons per cluster. The picking is required as some CoreAddons are specific to the type of cluster (e.g. chosen cloud provider, chosen CNI,...). Additionally the CoreAddonController protects against accidental deletion of AddonInstallation CRs of type Core. We think this makes sense as losing a CoreAddon (e.g. OpenVPN) could be detrimental to User Cluster health.
+b) Create CRs directly during cluster creation
 
-In order to know which UserCluster requires which CoreAddons, the CoreAddonController  works by maintaining a sub-list of CoreAddons for each UserCluster computed from the following sources:
+we would extend the `CreateEndpoint` func to:
 
-- cluster.kubermatic.io/v1 CR → Will be used for CoreAddons that depend on a clusters configuration (e.g. cilium vs canal). The logic for this will be coded directly into CoreAddonController as it can become quite complicated
-- CustomCoreAddon configmap →A configmap that contains a list of addons that should be added to every cluster. This map can be edited by the Cluster Admin. This allows Cluster Admins to specify their own Addons that should be shipped with every cluster, which is in line with the current [defaultManifests](https://docs.kubermatic.com/kubermatic/v2.18/guides/addons/) functionality
+- create the seed-namespace right after the cluster object has been created. We need to make sure this early creation does not interfere with the seed-controller
+- create the application CRs after the namespace has been created
 
-In order to protect against accidental deletion of an AddonInstallation of type Core, we propose to make use of k8s [finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/). In this case the CoreAddonController would automatically add itself as a finalizer to any AddonInstallation it creates. Then whenever an AddonInstallation is being deleted, kubernetes will wait with the deletion until the finalizer is removed. Afterwards CoreAddonController is going to run a couple of checks internally:
+This option would keep the cluster-object in its current size. However we are currently not sure if there are any implications from creating the namespace early. We think this will have no implication for cluster templates, as they are based on the selected UI values, which would contain the Application information.
```

Comment:
In fact, it may have an impact because the cluster template is stored in `clustertemplate` and cluster are created from `clustertemplateinstances` CR.  So I have to check if we may encounter the same kind of problem regarding the size.

Reply:
updated section with a new idea @vgramer and I developed